bpf/lib/nat.h

```diff
@@ -1764,27 +1762,86 @@ static __always_inline void snat_v6_init_tuple(const struct ipv6hdr *ip6,
 }
 
 static __always_inline bool snat_v6_needed(struct __ctx_buff *ctx,
-					   const union v6addr *addr)
+					   union v6addr *addr)
```

Comment:
Did you check whether this is still in sync with `snat_v4_prepare_state()` ? Right now it's a bit difficult to check whether we have all the corresponding cases covered.

I expected them to look slightly different because of EgressGW, but not sure that explains the whole diff :).

Reply:
I did check. I came up with [a commit](https://github.com/cilium/cilium/commit/243534ea087b84e08c09df3ab10380d31f76df9e) to update `snat_v6_needed()` with respects to `snat_v4_prepare_state()`, but so far I've not included it in the PR.

I noted that the function evolved mostly with the following commits:

  - cdfc30d ("bpf: egressgw: sync logic to determine if destination is outside cluster")
    Moving the `ct_is_reply()` check above the `IPV4_SNAT_EXCLUSION_DST_CIDR` logics makes sense to me, so we can probably apply the non-EGW part of this commit to `snat_v6_needed()`.

  - https://github.com/cilium/cilium/commit/453299601c8c5747deb0b004325db796a2fd31f9 ("bpf: nat: always track egress gateway connections")
    This one is some refactoring which is not necessary at the moment for IPv6, but it seems trivial to apply the few bits touching the relevant function to align with IPv4.

  - https://github.com/cilium/cilium/commit/59ac84f65d2c14fdf8038b29a53b2d1a87f18598 ("bpf: nat: consolidate snat_v4_needed() state into ipv4_nat_target")
    Same.

I probably missed some older commits in the list, but https://github.com/cilium/cilium/commit/243534ea087b84e08c09df3ab10380d31f76df9e brings the two versions (IPv4 and IPv6) pretty close to one another. I'm happy to rework it and add it to the PR.

Let's maybe sync sometimes on Slack/Zoom?

Reply:
Ah nice, thank you! I would need to stare a bit more at your current code to understand if something is genuinely missing, not sure how quickly we could fit in that session :/.

So if you already lined up a sync-up commit for a follow-on, I would probably be fine with merging as-is. And then let's review separately if the sync-up fixes any issues (and needs to go into v1.14), or can wait until v1.15.

Reply:
Alright, let's keep the current version and make it on par with IPv4 as a follow-up. I'll file an issue to list the different follow-ups to do for this PR.

Reply:
https://github.com/cilium/cilium/issues/26074